### ISSUE TITLE
IDate seems to be slower than Date

### ISSUE THREAD

**Issue by franknarf1** [Rdatatable/data.table#1836]
Example data

```
library(data.table)
set.seed(1)
id <- 1:1e4
date_samp <- seq(as.IDate("2010-01-01"),as.IDate("2011-01-01"),"day")
n_samp <- 5e7
DT <- data.table(
  id = sample(id,size = n_samp, replace=TRUE),
  date_1 = sample(date_samp,size = n_samp, replace=TRUE)
)
setkey(DT,id,date_1)

# also with integers
DT[, idt := as.integer(date_1)]
# also with vanilla dates
DT[, dt := as.Date(date_1)]
```

(Don't ask me why the data is keyed; I'm just following the lead of this SO post: http://stackoverflow.com/a/39283484/ )

Attempts

```
# the obvious way with IDate
system.time(DT[, diff := date_1 - shift(date_1), by=id]) # 25.90 seconds

# the obvious way with Date
system.time(DT[, diff2 := dt - shift(dt), by=id]) # 1.92

# other ways
system.time(DT[, 
  r := idt - shift(idt), by=id]) # 0.69

system.time(DT[, 
  r2 := idt - shift(idt)][
   idt != shift(idt), r2 := NA_integer_]) # 0.61
```

So IDate is a lot slower than Date for this fairly common task, it looks like.

(EDIT: Oops, was accidentally making difftimes into dates there. No effect on the timing.)

**Comment by MichaelChirico** [comment]
may be related to 1695? since there's no `-.IDate` it will default to using `-.Date`? seems like a crazy slowdown to be solely ascribed to that. I'll investigate a bit...

**Comment by MichaelChirico** [comment]
Minimal version of issue:

```
set.seed(1)
DT <- data.table(idate = as.IDate(sample(300, 5e7, TRUE), origin = "1970-01-01"))
system.time(DT[ , date := as.Date(idate)])
#    user  system elapsed 
#   0.176   0.072   0.248 

system.time(DT[ , as.POSIXct(idate)])
#    user  system elapsed 
#   4.936   0.316   5.247
system.time(DT[ , as.POSIXct(date)])
#    user  system elapsed 
#   0.180   0.056   0.234 
```

**Comment by jangorecki** [comment]
Probably just to create better posixct precision when having ITime on hand.

**Comment by MichaelChirico** [comment]
@jangorecki hmm I see.

The cost of that added precision is pretty evident until this goes internal... comparing to `as.POSIXct.Date`:

```
function (x, ...) 
.POSIXct(unclass(x) * 86400)
```

**Comment by mattdowle** [comment]
Looks like this is fixed now in the CRAN release.  @franknarf1 can you confirm this issue can be closed?  I ran your example with v1.10.4 and speed looks ok.

It seems I accidentally fixed it in [this commit](https://github.com/Rdatatable/data.table/commit/908be890f9e99f0f51a27eb42f2014ca853c7bac#diff-7347fe5a0f184f79ef064e92e3beb297) in Sep 2016 when releasing to CRAN then to make it pass another package's tests. @MichaelChirico also fixed it in [PR#1839](https://github.com/Rdatatable/data.table/pull/1839) which can be removed now?

Does the SO answer need an update?

**Comment by franknarf1** [comment]
> Looks like this is fixed now in the CRAN release. @franknarf1 can you confirm this issue can be closed? I ran your example with v1.10.4 and speed looks ok.

Yep, I'll close.

> Does the SO answer need an update?

I'll comment to the answerer/OP, pointing them to my new results (below), in case they want to update. It's fine if they do not update, I think.

```
> set.seed(1)
> id <- 1:1e4
> date_samp <- seq(as.IDate("2010-01-01"),as.IDate("2011-01-01"),"day")
> n_samp <- 5e7
> DT <- data.table(
+   id = sample(id,size = n_samp, replace=TRUE),
+   date_1 = sample(date_samp,size = n_samp, replace=TRUE)
+ )
> setkey(DT,id,date_1)
> 
> # also with integers
> DT[, idt := as.integer(date_1)]
> # also with vanilla dates
> DT[, dt := as.Date(date_1)]
> # the obvious way with IDate
> system.time(DT[, diff := date_1 - shift(date_1), by=id]) # 25.90 seconds
   user  system elapsed 
   0.58    0.06    0.64 
> 
> # the obvious way with Date
> system.time(DT[, diff2 := dt - shift(dt), by=id]) # 1.92
   user  system elapsed 
   1.63    0.07    1.69 
> 
> # other ways
> system.time(DT[, 
+   r := idt - shift(idt), by=id]) # 0.69
   user  system elapsed 
   0.50    0.01    0.52 
> 
> system.time(DT[, 
+   r2 := idt - shift(idt)][
+    idt != shift(idt), r2 := NA_integer_]) # 0.61
   user  system elapsed 
   0.30    0.19    0.48 
```

**Comment by mattdowle** [comment]
Great. Thanks @franknarf1.
